nx/lib/nx.ex

```diff
@@ -7142,6 +7142,42 @@ defmodule Nx do
     impl!(tensor).sort(tensor, tensor, axis: axis, comparator: comparator)
   end
 
+  @doc """
+    Solve the equation a x = b for x, assuming a is a triangular matrix.
+
+  # Examples
+
+  iex> Nx.triangular_solve([[3, 0, 0, 0], [2, 1, 0, 0], [1, 0, 1, 0], [1, 1, 1, 1]], [4, 2, 4, 2])
+  #Nx.Tensor<
+    f64[4]
+    [1.33333333, -0.66666667, 2.66666667, -1.33333333]
+  >
+
+  ### Error cases
+
+  iex> Nx.triangular_solve([[3, 0, 0, 0], [2, 1, 0, 0], [1, 0, 1, 0], [1, 1, 1, 1]], [4, 2, 4, 2], trans: 9)
+  ** (ArgumentError) unknown trans 9, expected 0 or 'N' for a x = b
+
+  """
+  @doc type: :linalg
+  def triangular_solve(a, b, opts \\ []) do
+    a = tensor(a)
+    b = tensor(b)
```

Comment:
```suggestion
    a = tensor!(a)
    b = tensor!(b)
```

`tensor!` is a bit more strict (it doesn't accept lists) and more optimized.

Reply:
@josevalim Thanks, also is there a way I can check if a tensor is square? 

I know I can use Nx.shape/1 but I am not sure how to actually validate any square matrix.

Reply:
See the eye/2 function, it will have an example.

Reply:
@josevalim when using `%T{shape: s1 = {m, n}} = a = tensor!(a)`, I get the following error: expected exception ArgumentError but got FunctionClauseError with message "no function clause matching in Nx.tensor!/1"

I am not sure what the issue here is.

Reply:
@AlexJuca your test cases are lists, they need to be Tensors that's why you're getting the error

Reply:
> @AlexJuca your test cases are lists, they need to be Tensors that's why you're getting the error

I totally missed that. Thanks for the heads up.